x-pack/plugin/apm-integration/src/main/java/org/elasticsearch/xpack/apm/APMTracer.java

```diff
@@ -103,56 +138,62 @@ protected void doStart() {
             .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
             .build();
-        tracer = openTelemetry.getTracer("elasticsearch", Version.CURRENT.toString());
+        var tracer = openTelemetry.getTracer("elasticsearch", Version.CURRENT.toString());
+
+        this.services = new APMServices(provider, tracer, openTelemetry);
     }
 
     @Override
     protected void doStop() {
-        final SdkTracerProvider provider = this.provider;
-        if (provider != null) {
-            provider.shutdown().join(30L, TimeUnit.SECONDS);
+        var services = this.services;
+        this.services = null;
+        if (services != null) {
+            services.provider.shutdown().join(30L, TimeUnit.SECONDS);
         }
     }
 
     @Override
-    protected void doClose() {}
+    protected void doClose() {
+
+    }
 
     @Override
     public void onTraceStarted(TracingPlugin.Traceable traceable) {
-        final Tracer tracer = this.tracer;
-        final OpenTelemetry openTelemetry = this.openTelemetry;
-        if (openTelemetry != null && tracer != null) {
-            spans.computeIfAbsent(traceable.getSpanId(), spanId -> {
-                final SpanBuilder spanBuilder = tracer.spanBuilder(traceable.getSpanName());
-                Context parentContext = getParentSpanContext(openTelemetry);
-                if (parentContext != null) {
-                    spanBuilder.setParent(parentContext);
-                }
-                for (Map.Entry<String, Object> entry : traceable.getAttributes().entrySet()) {
-                    final Object value = entry.getValue();
-                    if (value instanceof String) {
-                        spanBuilder.setAttribute(entry.getKey(), (String) value);
-                    } else if (value instanceof Long) {
-                        spanBuilder.setAttribute(entry.getKey(), (Long) value);
-                    } else if (value instanceof Integer) {
-                        spanBuilder.setAttribute(entry.getKey(), (Integer) value);
-                    } else if (value instanceof Double) {
-                        spanBuilder.setAttribute(entry.getKey(), (Double) value);
-                    } else if (value instanceof Boolean) {
-                        spanBuilder.setAttribute(entry.getKey(), (Boolean) value);
-                    } else {
-                        throw new IllegalArgumentException(
-                            "span attributes do not support value type of [" + value.getClass().getCanonicalName() + "]"
-                        );
-                    }
-                }
-                return spanBuilder.startSpan();
-            });
+        var services = this.services;
+        if (services == null) {
+            return;
         }
+
+        spans.computeIfAbsent(traceable.getSpanId(), spanId -> {
+            final SpanBuilder spanBuilder = services.tracer.spanBuilder(traceable.getSpanName());
```

Comment:
`services` could already be shut down at this point, if the shutdown happened after we read the field but before reaching this line.

Reply:
Correct. `spanBuilder` implementation has a check for this and will return noop

Reply:
Ok great. Could we add a comment saying that?